src/lib.rs

```diff
@@ -30,77 +30,91 @@ use criterion::{
-    target_arch = "aarch64",
-    target_arch = "loongarch64"
-)))]
-compile_error!(
-    "criterion-cycles-per-byte currently works only on x86 or x86_64 or aarch64 or loongarch64."
-);
-
 /// `CyclesPerByte` measures clock cycles using the CPU read time-stamp counter instruction. `cpb` is
 /// the preferred measurement for cryptographic algorithms.
 pub struct CyclesPerByte;
 
 // WARN: does not check for the cpu feature; but we'd panic anyway so...
-fn rdtsc() -> u64 {
+#[inline(always)]
+fn cycle_counter() -> u64 {
     #[cfg(target_arch = "x86_64")]
-    unsafe {
-        core::arch::x86_64::_mm_mfence();
-        core::arch::x86_64::_rdtsc()
-    }
-
+    use core::arch::x86_64::*;
     #[cfg(target_arch = "x86")]
+    use core::arch::x86::*;
+    
     unsafe {
-        core::arch::x86::_mm_mfence();
-        core::arch::x86::_rdtsc()
-    }
-
-    #[cfg(all(target_arch = "aarch64", target_os = "linux"))]
-    unsafe {
-        // If a aarch64 CPU, running GNU/Linux kernel, executes following instruction,
-        // it'll *probably* panic with message "illegal instruction executed", because userspace
-        // isn't allowed to execute that instruction without installing a Linux Kernel Module.
-        //
-        // I've tested the LKM @ https://github.com/jerinjacobk/armv8_pmu_cycle_counter_el0
-        // on a Raspberry Pi 4b ( i.e. ARM Cortex-A72, running kernel version 6.5.0-1006-raspi )
-        // and it works like charm. While extending support of this library for aarch64 targets,
-        // I found https://github.com/pornin/crrl#benchmarks pretty helpful.
-        let mut counter: u64;
-        core::arch::asm!("dsb sy", "mrs {}, pmccntr_el0", out(reg) counter);
-        counter
-    }
-
-    #[cfg(target_arch = "loongarch64")]
-    unsafe {
-        let counter: u64;
-        core::arch::asm!("rdtime.d {0}, $zero", out(reg) counter);
-        counter
+        cfg_if::cfg_if! {
+            if #[cfg(all(rdpru, any(target_arch = "x86_64", target_arch = "x86")))] {
+                let [hi, lo]: [u32; 2];
+                _mm_lfence();
+                core::arch::asm!(
+                    "rdpru",
+                    out("edx") hi,
+                    out("eax") lo,
+                    in("ecx") 1u32,
+                );
+                _mm_lfence();
+                (u64::from(hi) << 32) | u64::from(lo)
+            } else if #[cfg(any(target_arch = "x86", target_arch = "x86_64"))] {
+                // https://www.felixcloutier.com/x86/rdtscp
+                // - If software requires RDTSCP to be executed only after all previous stores are globally visible,
+                //   it can execute MFENCE immediately before RDTSCP.
+                // - If software requires RDTSCP to be executed prior to execution of any subsequent instruction
+                //   (including any memory accesses), it can execute LFENCE immediately after RDTSCP.
+                _mm_mfence();
+                let ret = __rdtscp();
```

Comment:
Note that I have changes RDTSC to RDTSCP. Strictly speaking, the former requires an additional LFENCE instruction before executing it as noted in https://www.felixcloutier.com/x86/rdtsc

Reply:
I reverted it back to RDTSC since RDTSCP may not be available on older CPUs. Plus, MFENCE is no longer needed with RDTSC as described in the added comment.